libraries/chain/controller.cpp

```diff
@@ -1602,6 +1602,17 @@ struct controller_impl {
       pending->push();
    }
 
+   void assign_signatures(block_state_ptr bsp, std::vector<signature_type>&& sigs, bool wtmsig_enabled) {
```

Comment:
I don't like this does more than assign signatures. I think it needs to be renamed or the `log_irreversible()` and `emit` should be pulled out of it. 
I also don't like that the `emit(self.accepted_block, bsp);` is called from two locations now. Seems we need a concept like finalize block (although we already have that). Maybe complete_block.